### ISSUE TITLE
[Task]: Model Contract - Foundation Types — Base Contracts, Data Types & Enums

### ISSUE THREAD

**Issue by rajeshgangireddy** [open-edge-platform/geti-instant-learn#1041]
### Description
Implement the core type system for the backend-agnostic model contract per RFC 01-model-contract.md.

#### 1. Enums & Errors (`library/src/instantlearn/utils/constants.py`)

Update the existing `Backend` enum and add new enums:

- `Backend(StrEnum)`: keep existing values + add `TORCH = "torch"` (alias for PYTORCH if needed)
- New: `PromptType(StrEnum)`: TEXT, VISUAL, POINT, BOX
- New: `ShotMode(StrEnum)`: ZERO_SHOT, ONE_SHOT, FEW_SHOT
- New: `ModelNotFittedError(RuntimeError)` — raised when `predict()` called before `fit()` on models that require it

#### 2. ModelCard (`library/src/instantlearn/models/model_card.py`) — NEW file

```python
@dataclass(frozen=True)
class ModelCard:
    name: str
    family: str
    description: str
    prompt_types: frozenset[PromptType]
    shot_modes: frozenset[ShotMode]
    exportable_to: frozenset[Backend]
```

#### 3. Sample & TensorSample (`library/src/instantlearn/data/base/sample.py`) — REPLACE

Numpy-only `Sample` dataclass:

- Fields: `image: np.ndarray | None` (HWC uint8/float32), `image_path: str | None`, `masks: np.ndarray | None` (N,H,W), `bboxes: np.ndarray | None` (N,4 xyxy), `points: np.ndarray | None` (N,K,2), `scores: np.ndarray | None` (N,), `categories: list[str]`, `category_ids: list[int] | None`
- Method: `to_tensors(device: str = "cpu") → TensorSample` (lazy torch import)
- Keep: `filter_by_category()` convenience method
- Drop: `is_reference`, `n_shot`, `mask_paths` (not in RFC)

`TensorSample` dataclass (torch counterpart): image (CHW), masks, bboxes, points, scores, categories, category_ids as tensors.

#### 4. Prediction (`library/src/instantlearn/data/base/prediction.py`) — NEW file

```python
@dataclass
class Prediction:
    masks: np.ndarray              # (N, H, W) bool/uint8
    scores: np.ndarray             # (N,) float32
    label_ids: np.ndarray          # (N,) int32
    label_names: np.ndarray        # (N,) str/object
    boxes: np.ndarray | None = None
    points: np.ndarray | None = None
    metadata: dict = field(default_factory=dict)

    def to_tensors(self, device: str = "cpu") -> dict[str, torch.Tensor]: ...
```

Not frozen — postprocessor pipeline mutates arrays in-place.

#### 5. Model ABC (`library/src/instantlearn/models/base.py`) — REPLACE

```python
class Model(ABC):
    @classmethod
    @abstractmethod
    def card(cls) -> ModelCard: ...

    @property
    @abstractmethod
    def backend(self) -> Backend: ...

    @abstractmethod
    def fit(self, reference: Sample | list[Sample]) -> None: ...

    @abstractmethod
    def predict(self, target: Sample | list[Sample]) -> list[Prediction]: ...
```

Pure ABC — zero torch imports at module level.

#### 6. TorchModel (`library/src/instantlearn/models/torch_base.py`) — NEW file

```python
class TorchModel(Model, nn.Module):
    def __init__(self, device, precision, preprocessor, postprocessor): ...

    @property
    def backend(self) -> Backend: return Backend.TORCH

    @abstractmethod
    def export(self, path: Path) -> Path: ...

    @abstractmethod
    def to_openvino(self) -> Model: ...

    def _prepare(self, target: Sample | list[Sample]) -> list[TensorSample]: ...
```

#### 7. OpenVINOModel (`library/src/instantlearn/models/openvino_base.py`) — NEW file

```python
class OpenVINOModel(Model):
    def __init__(self, model_dir, device="AUTO", preprocessor=None, postprocessor=None): ...

    @property
    def backend(self) -> Backend: return Backend.OPENVINO
```

No `from_pretrained()` on base — only models that support HF loading declare their own.

### Files to create/modify

| File | Action |
|------|--------|
| `library/src/instantlearn/utils/constants.py` | Modify (add enums) |
| `library/src/instantlearn/models/model_card.py` | Create |
| `library/src/instantlearn/data/base/sample.py` | Replace |
| `library/src/instantlearn/data/base/prediction.py` | Create |
| `library/src/instantlearn/models/base.py` | Replace |
| `library/src/instantlearn/models/torch_base.py` | Create |
| `library/src/instantlearn/models/openvino_base.py` | Create |
| `library/src/instantlearn/data/base/__init__.py` | Update exports |
| `library/src/instantlearn/models/__init__.py` | Update exports |

### Acceptance criteria

- [ ] All new files pass `ruff check` and `ruff format`
- [ ] `from instantlearn.models.base import Model` imports a torch-free ABC
- [ ] `from instantlearn.models.torch_base import TorchModel` works
- [ ] `from instantlearn.models.openvino_base import OpenVINOModel` works without torch installed
- [ ] `from instantlearn.data.base import Sample, Prediction` works without torch
- [ ] `Sample.to_tensors()` lazy-imports torch only when called

### Notes

- Fine to break things - and have not working library
- Internal components (prompt generators, encoders, SAM decoder) stay as `nn.Module`.
- Reference: https://github.com/open-edge-platform/geti-instant-learn/pull/1027/changes

**Comment by piotrgrubicki** [comment]
A new opened (#1057) because this one wasn't visible on the board.
